src/memory_service/fam_memory_service_direct.cpp

```diff
@@ -446,7 +490,133 @@ void Fam_Memory_Service_Direct::put_atomic(uint64_t regionId,
+        }
+    }
+    MEMORY_SERVICE_DIRECT_PROFILE_END_OPS(mem_direct_gather_strided_atomic)
+}
+
+void Fam_Memory_Service_Direct::scatter_indexed_atomic(
+    uint64_t regionId, uint64_t offset, uint64_t nElements,
+    const void *elementIndex, uint64_t elementSize, uint64_t key,
+    const char *nodeAddr, uint32_t nodeAddrSize) {
+    MEMORY_SERVICE_DIRECT_PROFILE_START_OPS()
+    ostringstream message;
+    string hashStr = "";
+    if (numAtomicThreads <= 0) {
+        message << "Atomic Transfer Library is not enabled";
+        throw Memory_Service_Exception(ATL_NOT_ENABLED, message.str().c_str());
+    }
+    hash<string> mystdhash;
+    hashStr = hashStr + std::to_string(regionId);
+    hashStr = hashStr + std::to_string(offset);
+    uint64_t qId = mystdhash(hashStr) % numAtomicThreads;
+    atomicMsg InpMsg;
+    memset(&InpMsg, 0, sizeof(atomicMsg));
+    InpMsg.nodeAddrSize = nodeAddrSize;
+    memcpy(&InpMsg.nodeAddr, nodeAddr, nodeAddrSize);
+    InpMsg.dstDataGdesc.regionId = regionId;
+    InpMsg.dstDataGdesc.offset = offset;
+    InpMsg.inElements = nElements;
+    InpMsg.indexSize = elementSize;
+    InpMsg.key = key;
+    InpMsg.flag |= ATOMIC_SCATTER_INDEX;
+
+    int ret = atomicQ[qId].push(&InpMsg, elementIndex);
+    if (ret) {
+        if (ret == ATL_QUEUE_FULL) {
+            message << "Atomic queue Full - Failed to insert";
+            throw Memory_Service_Exception(ATL_QUEUE_FULL,
+                                           message.str().c_str());
+        } else {
+            message << "Error inserting into Queue";
+            throw Memory_Service_Exception(ATL_QUEUE_INSERT_ERROR,
+                                           message.str().c_str());
+        }
+    }
+    MEMORY_SERVICE_DIRECT_PROFILE_END_OPS(mem_direct_scatter_indexed_atomic)
 }
+
+void Fam_Memory_Service_Direct::gather_indexed_atomic(
+    uint64_t regionId, uint64_t offset, uint64_t nElements,
+    const void *elementIndex, uint64_t elementSize, uint64_t key,
+    const char *nodeAddr, uint32_t nodeAddrSize) {
+    MEMORY_SERVICE_DIRECT_PROFILE_START_OPS()
+    ostringstream message;
+    string hashStr = "";
+    if (numAtomicThreads <= 0) {
+        message << "Atomic Transfer Library is not enabled";
+        throw Memory_Service_Exception(ATL_NOT_ENABLED, message.str().c_str());
+    }
+    hash<string> mystdhash;
+    hashStr = hashStr + std::to_string(regionId);
+    hashStr = hashStr + std::to_string(offset);
+    uint64_t qId = mystdhash(hashStr) % numAtomicThreads;
+    atomicMsg InpMsg;
+    memset(&InpMsg, 0, sizeof(atomicMsg));
+    InpMsg.nodeAddrSize = nodeAddrSize;
+    memcpy(&InpMsg.nodeAddr, nodeAddr, nodeAddrSize);
+    InpMsg.dstDataGdesc.regionId = regionId;
+    InpMsg.dstDataGdesc.offset = offset;
+    InpMsg.inElements = nElements;
+    InpMsg.indexSize = elementSize;
+    InpMsg.key = key;
+    InpMsg.flag |= ATOMIC_GATHER_INDEX;
+
+    int ret = atomicQ[qId].push(&InpMsg, elementIndex);
```

Comment:
"elementIndex" is a local string pointer, coming from rpc request parameter. 
In the "push" function, hope we are copying the content of this string and not saving the string pointer.

Looks like you are doing this operation at line# 177-184 atomic_queue.cpp